### PR TITLE
Video : Add frame interval APIs

### PR DESCRIPTION
**Pull request by ngphibang**
Video devices usually supports changing frame rates. This PR adds set/get/enum frame intervals APIs to support this.
It also 

- adds an implementation of the new APIs in the video software generator
- adds code to the video capture sample as an example of using these APIs

### CODE REVIEW

File: include/zephyr/drivers/video.h [hunk 1]
```diff
@@ -118,6 +118,66 @@ struct video_buffer {
...
+ */
+struct video_frmival_enum {
+	/** frame interval index during enumeration. */
+	uint32_t index;
+	/** FourCC pixel format value (\ref video_pixel_formats). */
+	uint32_t pixelformat;
+	/** frame width in pixels. */
+	uint32_t width;
+	/** frame height in pixels. */
+	uint32_t height;
```
Comment (josuah) [inline include/zephyr/drivers/video.h:171]: I think I missed something: this struct seems redundant with [`struct video_format`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/video.h#L43-L58).

It seems like the extra frame interval types could be an extension to the existing `struct video_format` instead?

Then there would [no need to introduce an extra API for this](https://github.com/zephyrproject-rtos/zephyr/assets/55351021/97c7e01c-7084-460d-9073-e247410e92f8) (from [here](https://static.linaro.org/connect/san19/presentations/san19-503.pdf#page=10))

In hope I am not mixing unrelated things...

Thank you very much for adding this essential point! Twice the FPS requires twice the hardware bandwidth, usually it is only possible at lower resolution.

Comment (ngphibang) [reply include/zephyr/drivers/video.h:171]: @josuah : Thanks. Fframe rate depends on pixel format, width and height.
Unlike other settings such as contrast, awb, etc. which are simply controlled by only the camera sensor and so, could be set which a video_cid setting, frame rate (and also format) dedicate an API because it depends on other hardware along the camera pipeline and need a runtime accordance between them.

**Comment by ngphibang** [comment]
@loicpoulain : Hi Loic, could you please help to review this one ?

File: include/zephyr/drivers/video.h (hunk 1 above)
Comment (ngphibang) [reply include/zephyr/drivers/video.h:171]: > I also wonder: if using set_frmival(YUYV 640x480px at 60FPS) and then set_format(RGB8 1280x720px), the width and height could be left inconsistent with the frame rate/interval for some sensors? Or the driver would have to automatically pick the right FPS.
> 
> Or would the driver have to auto-select the frame rate so that the requested size can be achieved?

- set_frmival() does not set the format (e.g. width and height), it just set the fps (but you need to provide width, height, pixel format as input) so there will be no "inconsistent" with set_fmt().
- It's the driver's responsability to handle cases that the user's required fps not supported. It can pick the nearest supported fps for that format or simply return error code.
- set_frmival() should not replace set_format() as already, the two APIs do different things, moreover set_frmival() is optional while set_format() is mandatory.

Comment (josuah) [reply include/zephyr/drivers/video.h:171]: > set_frmival() does not set the format

Ah, this is what I missed! Setting the frame interval requires width and height to be set, but they are only hints, do not change the values.

I also missed that these are all V4L2 semantics here.
These are documented in the Linux kernel.

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/media/v4l2-subdev.h#n784
https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-subdev-g-frame-interval.html

> Drivers must not return an error solely because the requested interval doesn’t match the device capabilities. They must instead modify the interval to match what the hardware can provide.
> 
> Changing the frame interval shall never change the format. Changing the format, on the other hand, may change the frame interval.

That might need links on the doc, but that is another pull request maybe.

Everything is clear now. Thank you a lot for taking the time of explaining earlier!

Comment (ngphibang) [reply include/zephyr/drivers/video.h:171]: > Drivers must not return an error solely because the requested interval doesn’t match the device capabilities. They must instead modify the interval to match what the hardware can provide.
> 
> Changing the frame interval shall never change the format. Changing the format, on the other hand, may change the frame interval.

Yes, I will add these in the code comments to be used as doc. Thanks !

**Comment by ngphibang** [comment]
@loicpoulain

File: drivers/video/video_sw_generator.c [hunk 2]
```diff
@@ -261,6 +270,66 @@ static inline int video_sw_generator_set_ctrl(const struct device *dev, unsigned
...
+
+	while (fmts[i].pixelformat && (fmts[i].pixelformat != fie->pixelformat)) {
+		i++;
+	}
+
+	if ((i == ARRAY_SIZE(fmts)) || (fie->width > fmts[i].width_max) ||
+	    (fie->width < fmts[i].width_min) || (fie->height > fmts[i].height_max) ||
+	    (fie->height < fmts[i].height_min)) {
+		return -EINVAL;
+	}
```
Comment (josuah) [inline drivers/video/video_sw_generator.c:319]: FYI: there is an opportunity to reduce the drivers code size by moving these to [`drivers/video.h`](https://github.com/zephyrproject-rtos/zephyr/pull/72254/files#diff-70c4de6949b97c13ac893b4bfb16fbcc433593fbfa96ec8947bb5cb83c4d04bcR485)

File: samples/subsys/video/capture/src/main.c [hunk 3]
```diff
@@ -78,6 +81,28 @@ int main(void)
...
+	while (video_enum_frmival(video, VIDEO_EP_OUT, &fie) == 0) {
+		if (fie.type == VIDEO_FRMIVAL_TYPE_DISCRETE) {
+			printk("   %u/%u ", fie.discrete.numerator, fie.discrete.denominator);
+		} else {
+			printk("   [min = %u/%u; max = %u/%u; step = %u/%u]\n",
+			       fie.stepwise.min.numerator, fie.stepwise.min.denominator,
+			       fie.stepwise.max.numerator, fie.stepwise.max.denominator,
+			       fie.stepwise.step.numerator, fie.stepwise.step.denominator);
+		}
+		fie.index++;
```
Comment (josuah) [inline samples/subsys/video/capture/src/main.c:103]: FYI: there is an opportunity to reduce the drivers code size by moving these to [`drivers/video.h`](https://github.com/zephyrproject-rtos/zephyr/pull/72254/files#diff-70c4de6949b97c13ac893b4bfb16fbcc433593fbfa96ec8947bb5cb83c4d04bcR485)
In the same style as [strsep](https://www.man7.org/linux/man-pages/man3/strsep.3.html).

File: include/zephyr/drivers/video.h [hunk 4]
```diff
@@ -318,6 +408,84 @@ static inline int video_get_format(const struct device *dev,
...
+		return -ENOSYS;
+	}
+
+	return api->get_frmival(dev, ep, frmival);
+}
+
+/**
+ * @brief List video frame intervals.
+ *
+ * List all supported video frame intervals of a given format.
```
Comment (josuah) [inline include/zephyr/drivers/video.h:466]: I like this API as it allows to lookup for other ~~`sensor`~~ ~~`sdev`~~ ~~`source_dev`~~ `source` devices to see what they support, although it is different from [`video_get_caps()`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/video.h#L466-L468). Just observing.

(reviewers: see [line 103](https://github.com/zephyrproject-rtos/zephyr/pull/72254/files#diff-5c275ef65ca2e3a0087625307d2de00a3c1edaa2d9edfa416ac4993a1755ffc5R103))

Would it help to comment how one is expected to walk through the enumeration? Stating that "`frmival` acts as a query" would also have helped with novice users (like me) to understand [this](https://github.com/zephyrproject-rtos/zephyr/pull/72254/files#r1590003732).

File: include/zephyr/drivers/video.h [hunk 5]
```diff
@@ -118,6 +118,66 @@ struct video_buffer {
...
 };
 
+/**
+ * @brief video_frmival_type enum
+ *
+ * Supported frame interval type of a video device.
+ */
+enum video_frmival_type {
+	VIDEO_FRMIVAL_TYPE_DISCRETE = 1,
+	VIDEO_FRMIVAL_TYPE_STEPWISE = 2,
```
Comment (josuah) [inline include/zephyr/drivers/video.h:128]: If `stepwise.min == stepwise.max` (and i.e. `stepwise.increment == 1` by convention), then `STEPWISE` becomes a `DISCRETE` value? This could be an opportunity to remove the distinction in the code and type definitions.

**Review by josuah — COMMENTED** [review]
Note: I am not a maintainer, not a collaborator, and not even a contributor yet!
Just free-form feedback... In hope not to bring more overhead.

**Comment by decsny** [comment]
@josuah thank you for helping review the video PRs, it needs more engagement in zephyr from people who know about it

File: samples/subsys/video/capture/src/main.c [hunk 6]
```diff
@@ -110,8 +135,8 @@ int main(void)
 			return 0;
 		}
 
-		printk("\rGot frame %u! size: %u; timestamp %u ms",
-		       frame++, vbuf->bytesused, vbuf->timestamp);
+		printk("\rGot frame %u! size: %u; timestamp %u ms\n", frame++, vbuf->bytesused,
```
Comment (loicpoulain) [inline samples/subsys/video/capture/src/main.c:138]: Unrelated change?

File: include/zephyr/drivers/video.h (hunk 5 above)
Comment (ngphibang) [reply include/zephyr/drivers/video.h:128]: I can't understand. If you remove STEP WISE, what will you describe the supported framerates if your driver supports all framerates from 1 to 1000 fps ? You will store 1000 values in an array ?

Comment (josuah) [reply include/zephyr/drivers/video.h:128]: I should have precised my thoughts better.
I thought about only keeping STEPWISE and using STEPWISE for DISCRETE values, rather than the other way around.

File: drivers/video/video_sw_generator.c (hunk 2 above)
Comment (ngphibang) [reply drivers/video/video_sw_generator.c:319]: How do you get "fmts" in video.h ?

Comment (josuah) [reply drivers/video/video_sw_generator.c:319]: You are right! Apologies I missed that there was no width/height as part of `fie`.
Each driver might encode that differently.
It could be added to something like `video_sensor_skeleton.c` (if that ever happens).
Sorry for the noise!

**Comment by loicpoulain** [comment]
@ngphibang why it could not be added as a generic control (set/get_ctrl)?

File: samples/subsys/video/capture/src/main.c (hunk 6 above)
Comment (ngphibang) [reply samples/subsys/video/capture/src/main.c:138]: It disturbed me a bit to view the logs so I fixed it as well but yes it should be in another commit, will do

**Comment by ngphibang** [comment]
@loicpoulain
Get/Set_Ctrl with VIDEO_CID_ are used for "simple" settings like contrast, brightness, 3A, etc. which concerns only the camera sensor, I think.
  
It's the same reason that we don't use set/get_ctrl for format I think. Framerate changing needs an accordance between elements in a camera pipeline (sensor, mipi csi receiver, etc.). It also depends on the capacity of related elements (that's why we need video_enum_frmival() as well). It also depends on which format chosen at runtime, so somehow a bit more complex than set/get_format.

File: include/zephyr/drivers/video.h (hunk 5 above)
Comment (ngphibang) [reply include/zephyr/drivers/video.h:128]: Sorry, but I don't see we could save code space here with your proposal. You save one structure in video.h by removing the DISCRETE type but in N drivers code, you waste more code space because STEP_WISE need more fields compared to DISCRETE (e.g., min, max, step).

A part from that, if I have a sensor that supports only 15, 20, 60 fps, I would describe it with a DISCRETE type with 3 values which are simple and "natural" than saying that it supports 3 STEP_WISE ranges:

- the 1st range is from 15 to 15 with step = 1
- the 2nd range is from 20 to 20 with step = 1
- the 3rd range is from 60 to 60 with step = 1

File: samples/subsys/video/capture/src/main.c (hunk 3 above)
Comment (ngphibang) [reply samples/subsys/video/capture/src/main.c:103]: I will look at this but it would be better if you could give a bit more detail

**Review by kartben — DISMISSED** [review]
some documentation related comments

**Comment by dleach02** [comment]
@ngphibang can you address @kartben change requests

**Comment by josuah** [comment]
> @ngphibang why it could not be added as a generic control (set/get_ctrl)?

As I understand it: the [`struct video_frmival_fie`](https://github.com/zephyrproject-rtos/zephyr/pull/72254/files#diff-70c4de6949b97c13ac893b4bfb16fbcc433593fbfa96ec8947bb5cb83c4d04bcR163-R179) is complex: "what frame rates are available for **this resolution** and **this format**".

One way to "squeeze" that multi-field query it through CIDs would be with several queries:
```c
video_set_ctrl(dev, VIDEO_CID_FRMIVAL_QUERY_WIDTH, 1280);
video_set_ctrl(dev, VIDEO_CID_FRMIVAL_QUERY_HEIGHT, 720);
video_set_ctrl(dev, VIDEO_CID_FRMIVAL_QUERY_FORMAT, video_fourcc(y, u, y, 2));
for (int i = 0;; i++) {
        video_set_ctrl(dev, VIDEO_CID_FRMIVAL_QUERY_IDX, i);
        if (video_get_ctrl(dev, VIDEO_CID_FRMIVAL_QUERY_IDX, &result)) {
                break;
        }
        /* do something with `struct result` */
}
```

File: include/zephyr/drivers/video.h (hunk 4 above)
Comment (ngphibang) [reply include/zephyr/drivers/video.h:473]: ok, I added some texts

File: include/zephyr/drivers/video.h [hunk 7]
```diff
@@ -118,6 +118,66 @@ struct video_buffer {
...
+
+/**
+ * @struct video_frmival
+ * @brief Video frame interval structure
+ *
+ * Used to describe a video frame interval.
+ */
+struct video_frmival {
+	uint32_t numerator;
+	uint32_t denominator;
```
[the remark this answers is not in the document]
Comment (ngphibang) [reply include/zephyr/drivers/video.h]: yes, fixed, thanks

File: include/zephyr/drivers/video.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (ngphibang) [reply include/zephyr/drivers/video.h]: yes, fixed, thanks

**Comment by dleach02** [comment]
@kartben please confirm if @ngphibang has addressed your change requests.

File: include/zephyr/drivers/video.h (hunk 5 above)
Comment (josuah) [reply include/zephyr/drivers/video.h]: Sorry for the very delayed answer, I missed your reply.
I understand now: the overhead of dealing with a cumbersome API ("a to a" instead of "a") is greater than just 1 saved struct.
Thank you!

Comment (ngphibang) [reply include/zephyr/drivers/video.h]: @josuah : Sorry if I misunderstand your last phrase because I am not sure if you agree or disagree with my explanation :) ?
In my previous comment, I mean that if we remove the DISCRETE type, we will save one struct in video.h, but it will be not natural to describe the discrete framerate (with the above example). Moreover, I think we will waste some code space in drivers because the STEP_WISE struct has more fields than the DISCRETE struct if we use it to describe the discrete framerates.

Comment (josuah) [reply include/zephyr/drivers/video.h]: I do agree with your explanation! :)
I understand the benefit of introducing an extra struct for the sake of making the configuration of drivers less tedious and confusing, and waste code space!

**Review by josuah — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

File: include/zephyr/drivers/video.h [hunk 8]
```diff
@@ -118,6 +118,70 @@ struct video_buffer {
...
+};
+
+/**
+ * @struct video_frmival_enum
+ * @brief Video frame interval enumeration structure
+ *
+ * Used to describe the supported video frame intervals of a given format.
+ */
+struct video_frmival_enum {
+	/** frame interval index during enumeration */
```
Comment (loicpoulain) [inline include/zephyr/drivers/video.h:168]: This seems to include what we already describe in video_format struct, why not using/passing that struct instead.

**Comment by ngphibang** [comment]
Rebased due to https://github.com/zephyrproject-rtos/zephyr/pull/73001 is merged which moved the video samples into a /drivers.

@loicpoulain : Could you please help to review this as it has been here for some months ?

For your question, as explained above, get/set_format as well as get/set_frmival cannot be done using get/set_ctrl for sure.

Semantically, get/set_ctrl is used for settings that relates only to the camera sensor, not the whole pipeline while format and framerate concern the whole pipeline. When you change the format or the frame rate, not only the sensor has to be changed, but only the receiver (e.g. mipi csi2rx) has to be changed to afford it (i.e., format / framerate changes imply to changes in pixel rate, link frequency, etc.

Technically, it needs to implemented as a pipeline API. For an example of a real usecase, [this PR](https://github.com/zephyrproject-rtos/zephyr/pull/76144) implements the frame rate changing for the csi -> mipi csi2rx -> ov5640 pipeline.

File: include/zephyr/drivers/video.h (hunk 8 above)
Comment (ngphibang) [reply include/zephyr/drivers/video.h:168]: @loicpoulain I 'd like to do so but unfortunately, the video_format struct contains the **_pitch field_** which is not used here. Unused fields waste memory space especially when it is in an API definition as many drivers will implement it.

Otherwise, I think the pitch field is not needed in the video_format struct as it can be computed separately from pixel format and width field. If you feel agree with this, I can submit a PR to remove this field from video_format then using the video_format struct here ?

Comment (loicpoulain) [reply include/zephyr/drivers/video.h:168]: Hmm AFAIU, pitch can depend on the capture pipeline, and depend also on some alignment requirements, so not sure it's something we can obtain only from pixeld format and width.

When we want to enumerate the allowed frame intervals, we normally already picked a frame format (e.g video_get_format), so it's odd to me we need to copy it's value back for enumerating frame intervals, instead of just passing a reference to the current format.

Comment (ngphibang) [reply include/zephyr/drivers/video.h:168]: Yes, I agree that pitch in the simplest case is` width * bpp` but in general it depends on other factors such as alignment.

BTW, for a clean code, especially for an API definition, I think we should not pass something that is more than needed, AFAIK in upstream Linux. It will create a bad precedent if we do it.

Comment (ngphibang) [reply include/zephyr/drivers/video.h:168]: @loicpoulain: Replacing :

```
struct video_frmival_enum {
	...
	uint32_t pixelformat;
	uint32_t width;
	uint32_t height;
	...
}
```
by 

```
struct video_frmival_enum {
	...
	struct video_format;
	...
}
```

where

```
struct video_format {
	uint32_t pixelformat;
	uint32_t width;
	uint32_t height;

	**_uint32_t pitch;_**
};
```

would be convenient if we already have the current video_format struct, so we just pass it to the function.
But if we want to query the supported frame rates for other video formats, we still have to construct the video_format struct anyways and leaving a never used field (e.g. pitch) in the struct will make developers confused as they don't know why they need or how to fill this pitch field, I think.

Other drawbacks is that never-used fields will waste memory space and could make a "dangerous point" (imagine that this field will be sent from application to the drivers without any sanity checks ...).

But well, if you see it's more convenient this way, I am going to change it.

**Review by dleach02 — DISMISSED** [review]

**Comment by dleach02** [comment]
@loicpoulain need your review on this PR. It is needed before a series of others that are pending.

File: include/zephyr/drivers/video.h (hunk 8 above)
Comment (josuah) [reply include/zephyr/drivers/video.h:168]: Maybe the pitch does have some usefulness...
It introduces the idea of padding, and the padding data, while not useful, affects the frame rate too?

For instance, these two rows of pixels would surely take different amount of time to transmit:

```
[  frame data  |  padding              ] <- larger pitch,same width
```
```
[  frame data  |  padding  ] <- smaller pitch, same width
```

That sounds like some rare use-case though, at least for uncompressed frame images.

Comment (josuah) [reply include/zephyr/drivers/video.h:168]: Now that `struct video_format` is there, would it work to pass it separately?

Something like this:
```diff
 static int video_xyzzy_enum_frmival(const struct device *dev, enum video_endpoint_id ep,
+				     const struct video_format *query,
				     struct video_frmival_enum *response);
```

Then the application could have a const (flash) array of video formats it supports, and try to pass them one by one?

```c
for (int i = 0; i < ARRAY_SIZE(supported_fmts); i++) {
	struct video_frmival_enum response;
	struct video_frmival frmival;

	video_enum_frmival(video_dev, VIDEO_EP_OUT, &supported_fmts[i], &response);
	if (search_acceptable_frame_interval(&response, &frmival) == 0) {
		return video_set_frmival(video_dev, VIDEO_EP_OUT, &frmival);
	}
}
return -ENOTSUP;
```

Maybe this is what @loicpoulain meant by "passing a reference".

**Comment by josuah** [comment]
Letting the conversation above get resolved. No issue on my end with this way of solving format negotiation. I better avoid open-ended discussions in the middle of reviews to not delay the "go".

**Review by josuah — DISMISSED** [review]

File: include/zephyr/drivers/video.h (hunk 8 above)
Comment (ngphibang) [reply include/zephyr/drivers/video.h:168]: @josuah : Sorry I was a bit late on responding ... Yes, I understood what @loicpoulain meant is to pass a `video_format` struct as a reference to the `enum_frmival()` API instead of passing 3 separate parameters (width, height, pixel_format). It's more convenient to do that as the `video_format `struct is already defined.

The point being discussed is the `video_format` struct contains a pitch field which we will never use. You are right that the `pitch ` does affect the pixel rate / frame rate (although it is usually small and neglectable). But the main reason here is to query of supported frame rates, we don't need to give the pitch. We ask the pipeline :

- Give me the frame rates supported for this `{width, height, pixel_format}`

The pipeline (e.g. : ov5640 -> mipi csi2rx -> csi) will know to answer with only these parameters as input. The sensor driver itself already knows which frame rates it can support for 720p RGB565 format for example, **if there are any padding needed, it already took them into account internally**. Then, the mipi csi2rx will base on the required pixel rate or link frequency from the sensor (see https://github.com/zephyrproject-rtos/zephyr/pull/76475/commits/e3ad02203ed5f57874071e735c635efa7b57065c#diff-f7423f09c35fbf23beefd1f706f4e1dbd5477daf2c40fbc1d6b731d206156bd9R239) to derive the supported frame rates of the whole pipeline.

It's like when you ask a carpenter "Give me the price of a wooden bookcase of size 1m x 0.5m x 2m", you don't need to precise "you should take a little bit more wood than 1x 0.5 x 2" to compensate for the loss during the production".

Comment (josuah) [reply include/zephyr/drivers/video.h:168]: Good point, no need to explain to a device what is its own pitch: it knows about it already.

File: samples/subsys/video/capture/src/main.c (hunk 3 above)
Comment (josuah) [reply samples/subsys/video/capture/src/main.c]: My bad for forgetting to answer.
I was thinking of auto-incrementing `fie.index` as part of `video_enum_frmival()` or something similar, but having an explicit `fie.index++;` helps with readability.

File: include/zephyr/drivers/video.h (hunk 8 above)
Comment (ngphibang) [reply include/zephyr/drivers/video.h:168]: @loicpoulain : Replacing `{width, height, pixel_format}` by `struct video_format` as your request (the pitch field is never used but well it's not a big deal).

**Comment by ngphibang** [comment]
@loicpoulain : Could you please revisit this PR to move it forward ?
Thanks !

**Review by loicpoulain — APPROVED** [review]

**Review by josuah — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]
